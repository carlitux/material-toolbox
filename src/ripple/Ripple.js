// @flow
import * as React from 'react';
import ReactDom from 'react-dom';
import classnames from 'classnames';

import * as util from '@material/ripple/util';
import MDCRippleFoundation from '@material/ripple/foundation';

type Arguments = {
  unbounded: boolean,
  disabled: boolean,
};

type State = {
  classes: { [string]: boolean },
};

const Ripple = (
  Component: React.ComponentType<any>,
  { unbounded, disabled }: Arguments = { unbounded: false, disabled: false },
) =>
  class extends React.Component<{ [any]: any, className: string }, State> {
    static MATCHES: string = util.getMatchesProperty(HTMLElement.prototype);
    displayName = `${Component.displayName || ''}Rippled`;
    state = { classes: {} };

    componentDidMount() {
      this.foundation.init();
    }

    componentWillUnmount() {
      this.foundation.destroy();
    }

    activate() {
      this.foundation.activate();
    }

    deactivate() {
      this.foundation.deactivate();
    }

    foundation = new MDCRippleFoundation({
      browserSupportsCssVars: () => util.supportsCssVariables(window),
      isUnbounded: () => unbounded,
      isSurfaceActive: () => this.element[this.constructor.MATCHES](':active'),
      isSurfaceDisabled: () => disabled,
      addClass: className =>
        this.element &&
        this.setState({
          classes: { ...this.state.classes, [className]: true },
        }),
      removeClass: className =>
        this.element &&
        this.setState({
          classes: { ...this.state.classes, [className]: false },
        }),
      registerInteractionHandler: (evtType, handler) =>
        this.element.addEventListener(evtType, handler, util.applyPassive()),
      deregisterInteractionHandler: (evtType, handler) =>
        this.element.removeEventListener(evtType, handler, util.applyPassive()),
      registerResizeHandler: handler =>
        window.addEventListener('resize', handler),
      deregisterResizeHandler: handler =>
        window.removeEventListener('resize', handler),
      updateCssVariable: (varName, value) =>
        this.element && this.element.style.setProperty(varName, value),
      computeBoundingRect: () => this.element.getBoundingClientRect(),
      getWindowPageOffset: () => ({
        x: window.pageXOffset,
        y: window.pageYOffset,
      }),
    });
    element: HTMLElement;

    render() {
      const className = classnames(this.state.classes, this.props.className);

      return (
        <Component
          {...this.props}
          className={className}
          rippleRef={element => {
            this.element = ReactDom.findDOMNode(element); // eslint-disable-line
          }}
        />
      );
    }
  };

export default Ripple;
