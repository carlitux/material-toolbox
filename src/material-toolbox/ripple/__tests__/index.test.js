// @flow
import React from 'react';
import { shallow, mount } from 'enzyme';

import Ripple from '../index';

describe('component::Ripple', () => {
  it('Should match snapshoot', () => {
    const component = mount(
      <Ripple>
        <div>Ripple</div>
      </Ripple>,
    );
    expect(component).toMatchSnapshot();
  });

  it('Should match snapshoot cssOnly', () => {
    const component = shallow(
      <Ripple cssOnly>
        <div>Ripple</div>
      </Ripple>,
    );
    expect(component).toMatchSnapshot();
  });

  it('Should match snapshoot primary', () => {
    const component = mount(
      <Ripple primary>
        <div>Ripple</div>
      </Ripple>,
    );
    expect(component).toMatchSnapshot();
  });

  it('Should match snapshoot accent', () => {
    const component = mount(
      <Ripple accent>
        <div>Ripple</div>
      </Ripple>,
    );
    expect(component).toMatchSnapshot();
  });
});
