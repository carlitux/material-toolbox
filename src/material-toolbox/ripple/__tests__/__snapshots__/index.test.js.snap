// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component::Ripple Should match snapshoot 1`] = `
<Ripple
  accent={false}
  cssOnly={false}
  disabled={false}
  primary={false}
  unbounded={false}
>
  <div
    className="mdc-ripple-surface"
    key=".0"
    style={Object {}}
  >
    Ripple
  </div>
</Ripple>
`;

exports[`component::Ripple Should match snapshoot accent 1`] = `
<Ripple
  accent={true}
  cssOnly={false}
  disabled={false}
  primary={false}
  unbounded={false}
>
  <div
    className="mdc-ripple-surface mdc-ripple-surface--accent"
    key=".0"
    style={Object {}}
  >
    Ripple
  </div>
</Ripple>
`;

exports[`component::Ripple Should match snapshoot cssOnly 1`] = `
Array [
  <div
    className="mdc-ripple-surface"
    key=".0"
    style={Object {}}
  />,
]
`;

exports[`component::Ripple Should match snapshoot primary 1`] = `
<Ripple
  accent={false}
  cssOnly={false}
  disabled={false}
  primary={true}
  unbounded={false}
>
  <div
    className="mdc-ripple-surface mdc-ripple-surface--primary"
    key=".0"
    style={Object {}}
  >
    Ripple
  </div>
</Ripple>
`;
